### ISSUE TITLE
ORA-12162 when connecting to database since upgrade to 1.3.0

### ISSUE THREAD

**Issue by glM26** [oracle/python-oracledb#178]
1. What versions are you using?

python-oracledb 1.3.1
Oracle DB 19c
RHEL 7.9
Python 3.8

2. Is it an error or a hang or a crash?
error
3. What error(s) or behavior you are seeing?
We are using python-oracledb in thick-mode to connect to an Oracle database specified in a tnsnames.ora file (i.e. TNS_ADMIN is set to point to the tnsnames.ora file).
We are also using a sqlnet.ora file that contains a reference to a wallet that contains the DB user and password.

We connect to the database in the following way:
```
>>> import oracledb;oracledb.init_oracle_client();oracledb.connect(dsn="TEST.WORLD")
<oracledb.Connection to externally identified user>
```

This is working in python-oracledb up to version 1.2.2.
Since version 1.3.0 (as well as 1.3.1) the same code now results in an error:
```
>>> import oracledb;oracledb.init_oracle_client();oracledb.connect(dsn="TEST.WORLD")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/PYTHON_LIBS_38/lib64/python3.8/site-packages/oracledb/connection.py", line 1008, in connect
    return conn_class(dsn=dsn, pool=pool, params=params, **kwargs)
  File "/PYTHON_LIBS_38/lib64/python3.8/site-packages/oracledb/connection.py", line 133, in __init__
    impl.connect(params_impl, pool_impl)
  File "src/oracledb/impl/thick/connection.pyx", line 387, in oracledb.thick_impl.ThickConnImpl.connect
  File "src/oracledb/impl/thick/utils.pyx", line 413, in oracledb.thick_impl._raise_from_odpi
  File "src/oracledb/impl/thick/utils.pyx", line 403, in oracledb.thick_impl._raise_from_info
oracledb.exceptions.DatabaseError: ORA-12162: TNS:net service name is incorrectly specified
```

The tnsnames.ora file:
```
TEST.WORLD =
  (DESCRIPTION =
    (CONNECT_TIMEOUT=10)(RETRY_COUNT=3)
    (ADDRESS_LIST =
      (ADDRESS=(PROTOCOL=TCP)(HOST= primary_url)(PORT=1721))
      (ADDRESS=(PROTOCOL=TCP)(HOST= secondary_url)(PORT=1721))
    )
    (CONNECT_DATA=(SERVICE_NAME=TEST_PRIM.WORLD)(SERVER=DEDICATED))
 )
```

I can still connect to the database in 1.3.1 with all parameters in the connect-statement:
```
>>> import oracledb;oracledb.init_oracle_client();oracledb.connect(host="primary_url", port=1721, service_name="TEST.WORLD", user="TEST", password="")

<oracledb.Connection to TEST@(DESCRIPTION=(ADDRESS=(PROTOCOL=tcp)(HOST=primary_url)(PORT=1721))(CONNECT_DATA=(SERVICE_NAME=TEST.WORLD)))>
```

Have there been any changes in release 1.3.0 that I overlooked or is there a bug in release 1.3.0?

**Comment by anthony-tuininga** [comment]
I'm not immediately aware of any changes that might be relevant. Looking at your sample tnsnames.ora I see the service_name is `TEST_PRIM.WORLD`, not `TEST.WORLD` as in your connection with all of the parameters. Is that just an oversight? I'll look into this and get back to you.

**Comment by glM26** [comment]
> I'm not immediately aware of any changes that might be relevant. Looking at your sample tnsnames.ora I see the service_name is `TEST_PRIM.WORLD`, not `TEST.WORLD` as in your connection with all of the parameters. Is that just an oversight? I'll look into this and get back to you.

We have been using the TEST_PRIM.WORLD in tnsnames.ora for quite some time now, while still connecting to TEST.WORLD.
I do not know why we use this, as we got the tnsnames.ora in this form from our DBA-team a couple of years ago.

But I just tried to change TEST_PRIM.WORLD to TEST.WORLD and it does not change the behavior, i.e. ORA-12162.

**Comment by cjbj** [comment]
@glM26 would you be able to create a sqlnet.ora file (on the machine where your tnsnames.ora file is) and add some tracing like:

**Comment by glM26** [comment]
@cjbj 
I added the trace-options to sqlnet.ora, it now looks like this:
```
WALLET_LOCATION =
    (SOURCE =
        (METHOD = FILE)
        (METHOD_DATA =
            (DIRECTORY = /opt/shared/shsdb/operation/etc/oracle_wallet)
        )
    )
SQLNET.WALLET_OVERRIDE = TRUE
TRACE_LEVEL_CLIENT=USER
ADR_BASE=/tmp/mylogs
DIAG_ADR_ENABLED=ON
```
With this I get the following trace-log (only first  64 lines) with oracledb 1.2.2:
```
Trace file /tmp/mylogs/oradiag_te504ho2/diag/clients/user_te504ho2/host_885234334_110/trace/ora_529677_140072190089088.trc
2023-05-09 07:26:05.034 : nlstddt_do_alter_trace:Version 19.3.0.0.0
2023-05-09 07:26:05.042 : nlstddt_do_alter_trace:--- TRACE CONFIGURATION INFORMATION FOLLOWS ---
2023-05-09 07:26:05.042 : nlstddt_do_alter_trace:New trace stream is /tmp/mylogs/oradiag_te504ho2/diag/clients/user_te504ho2/host_885234334_110/trace/ora_529677_140072190089088.trc
2023-05-09 07:26:05.042 : nlstddt_do_alter_trace:New trace level is 4
2023-05-09 07:26:05.042 : nlstddt_do_alter_trace:--- TRACE CONFIGURATION INFORMATION ENDS ---
2023-05-09 07:26:05.042 : nlstdtp_trace_pfile:--- PARAMETER SOURCE INFORMATION FOLLOWS ---
2023-05-09 07:26:05.042 : nlstdts_trace_source:Attempted load of system pfile source /opt/sas/share/ora/sqlnet.ora
2023-05-09 07:26:05.042 : nlstdts_trace_source:Parameter source loaded successfully
2023-05-09 07:26:05.042 : nlstdts_trace_source:
2023-05-09 07:26:05.042 : nlstdts_trace_source:Attempted load of local pfile source /home/in.bundesbank.de/te504ho2/.sqlnet.ora
2023-05-09 07:26:05.042 : nlstdts_trace_source:Parameter source was not loaded
2023-05-09 07:26:05.042 : nlstdts_trace_source:
2023-05-09 07:26:05.042 : nlstdtp_trace_pfile: -> PARAMETER TABLE LOAD RESULTS FOLLOW <-
2023-05-09 07:26:05.042 : nlstdtp_trace_pfile:Successful parameter table load
2023-05-09 07:26:05.042 : nlstddp_dump_ptable: -> PARAMETER TABLE HAS THE FOLLOWING CONTENTS <-
2023-05-09 07:26:05.042 : nlstddp_dump_ptable:  DIAG_ADR_ENABLED = ON
2023-05-09 07:26:05.042 : nlstddp_dump_ptable:  TRACE_LEVEL_CLIENT = USER
2023-05-09 07:26:05.042 : nlstddp_dump_ptable:  WALLET_LOCATION = (SOURCE = (METHOD = FILE) (METHOD_DATA = (DIRECTORY = /opt/shared/shsdb/operation/etc/oracle_wallet)))
2023-05-09 07:26:05.042 : nlstddp_dump_ptable:  ADR_BASE = /tmp/mylogs
2023-05-09 07:26:05.042 : nlstddp_dump_ptable:  SQLNET.WALLET_OVERRIDE = TRUE
2023-05-09 07:26:05.042 : nlstdtp_trace_pfile:--- PARAMETER SOURCE INFORMATION ENDS ---
2023-05-09 07:26:05.042 : nlstddl_do_alter_log:--- LOG CONFIGURATION INFORMATION FOLLOWS ---
2023-05-09 07:26:05.042 : nlstddl_do_alter_log:Log stream will be "/tmp/mylogs/oradiag_te504ho2/diag/clients/user_te504ho2/host_885234334_110/alert/log.xml"
2023-05-09 07:26:05.042 : nlstddl_do_alter_log:Log stream validation not requested
2023-05-09 07:26:05.042 : nlstddl_do_alter_log:--- LOG CONFIGURATION INFORMATION ENDS ---

2023-05-09 07:26:05.043 : nrigbni:Unable to get data from navigation file tnsnav.ora
2023-05-09 07:26:05.069 : nigconcbs:Can't Get Oracle SID Name
2023-05-09 07:26:13.370 : nnftmlf_make_local_addrfile:construction of local names file failed
2023-05-09 07:26:13.371 : nnftmlf_make_system_addrfile:system names file is /opt/sas/share/ora/tnsnames.ora
2023-05-09 07:26:13.373 : snldlldl:Couldn't load shared library nque
2023-05-09 07:26:13.373 : snldlldl:Err: /opt/oraclient/product/19.0.0/client3_1/lib/libnque19.so: undefined symbol: ssMemFree
2023-05-09 07:26:13.373 : niotns:Not trying to enable dead connection detection.
2023-05-09 07:26:13.373 : niotns:Calling address: (DESCRIPTION=(CONNECT_TIMEOUT=10)(RETRY_COUNT=3)(ADDRESS_LIST=(ADDRESS=(PROTOCOL=TCP)(HOST=host_name1)(PORT=1721))(ADDRESS=(PROTOCOL=TCP)(HOST=host_name2)(PORT=1721)))(CONNECT_DATA=(SERVICE_NAME=TEST_PRIM.WORLD)(SERVER=DEDICATED)(CID=(PROGRAM=python3)(HOST=origin_host)(USER=origin_os_user))))
2023-05-09 07:26:13.377 : nscall:connecting...
2023-05-09 07:26:13.380 : nsc2addr:(DESCRIPTION=(CONNECT_TIMEOUT=10)(RETRY_COUNT=3)(ADDRESS=(PROTOCOL=TCP)(HOST=10.204.8.23)(PORT=1721)(HOSTNAME=host_name1))(CONNECT_DATA=(SERVICE_NAME=TEST_PRIM.WORLD)(SERVER=DEDICATED)(CID=(PROGRAM=python3)(HOST=origin_host)(USER=origin_os_user))))
2023-05-09 07:26:13.380 : nttbnd2addr:using host IP address: 10.204.8.23
2023-05-09 07:26:13.380 : nstoSetupTimeout:ATO enabled for ctx=0x0x12dc4e0, val=10000(millisecs)
2023-05-09 07:26:13.380 : nstoUpdateActive:Active timeout is 0 (see nstotyp)
2023-05-09 07:26:13.381 : nsopen:opening transport...
2023-05-09 07:26:13.381 : nttcni:Tcp conn timeout = 60000 (ms)
2023-05-09 07:26:13.381 : nttcni:trying to connect to socket 9.
2023-05-09 07:26:13.381 : nstoControlTTO:Setup TTO: ATO=10000 TTO=60000
2023-05-09 07:26:13.381 : nsnainit:inf->nsinfflg[0]: 0x41 inf->nsinfflg[1]: 0x41
2023-05-09 07:26:13.381 : nsopen:global context check-in (to slot 0) complete
2023-05-09 07:26:13.381 : nscall:deferring connect attempt; at stage 104
2023-05-09 07:26:13.383 : nttctl:set TCP_NODELAY on 9
2023-05-09 07:26:13.383 : nscon:doing connect handshake...
2023-05-09 07:26:13.383 : nscon:connect id = 0xa4e4
2023-05-09 07:26:13.383 : nscon:sending NSPTCN packet
2023-05-09 07:26:13.383 : nscon:sending 286 bytes connect data
2023-05-09 07:26:13.383 : nsdo:286 bytes to NS buffer
2023-05-09 07:26:13.383 : nsevdcall:exit
2023-05-09 07:26:13.425 : nscon:got NSPTRS packet
2023-05-09 07:26:13.425 : nsevdcall:exit
2023-05-09 07:26:13.425 : nttctl:set TCP_NODELAY on 9
2023-05-09 07:26:13.425 : nscon:connect id = 0xa4e4
2023-05-09 07:26:13.425 : nscon:sending NSPTCN packet
2023-05-09 07:26:13.425 : nscon:sending 286 bytes connect data
2023-05-09 07:26:13.425 : nsdo:286 bytes to NS buffer
2023-05-09 07:26:13.425 : nsevdcall:exit
2023-05-09 07:26:13.426 : nscon:got NSPTAC packet
2023-05-09 07:26:13.426 : nscon:connect handshake is complete
```

When running with oracledb 1.3.0 it looks similar in the beginning, but then it just stops at `nigconcbs:Can't Get Oracle SID Name`:
```
Trace file /tmp/mylogs/oradiag_te504ho2/diag/clients/user_te504ho2/host_885234334_110/trace/ora_550243_140676284045184.trc
2023-05-09 07:45:46.259 : nlstddt_do_alter_trace:Version 19.3.0.0.0
2023-05-09 07:45:46.271 : nlstddt_do_alter_trace:--- TRACE CONFIGURATION INFORMATION FOLLOWS ---
2023-05-09 07:45:46.271 : nlstddt_do_alter_trace:New trace stream is /tmp/mylogs/oradiag_te504ho2/diag/clients/user_te504ho2/host_885234334_110/trace/ora_550243_140676284045184.trc
2023-05-09 07:45:46.271 : nlstddt_do_alter_trace:New trace level is 4
2023-05-09 07:45:46.271 : nlstddt_do_alter_trace:--- TRACE CONFIGURATION INFORMATION ENDS ---
2023-05-09 07:45:46.271 : nlstdtp_trace_pfile:--- PARAMETER SOURCE INFORMATION FOLLOWS ---
2023-05-09 07:45:46.271 : nlstdts_trace_source:Attempted load of system pfile source /opt/sas/share/ora/sqlnet.ora
2023-05-09 07:45:46.271 : nlstdts_trace_source:Parameter source loaded successfully
2023-05-09 07:45:46.271 : nlstdts_trace_source:
2023-05-09 07:45:46.271 : nlstdts_trace_source:Attempted load of local pfile source /home/in.bundesbank.de/te504ho2/.sqlnet.ora
2023-05-09 07:45:46.271 : nlstdts_trace_source:Parameter source was not loaded
2023-05-09 07:45:46.271 : nlstdts_trace_source:
2023-05-09 07:45:46.271 : nlstdtp_trace_pfile: -> PARAMETER TABLE LOAD RESULTS FOLLOW <-
2023-05-09 07:45:46.271 : nlstdtp_trace_pfile:Successful parameter table load
2023-05-09 07:45:46.271 : nlstddp_dump_ptable: -> PARAMETER TABLE HAS THE FOLLOWING CONTENTS <-
2023-05-09 07:45:46.271 : nlstddp_dump_ptable:  DIAG_ADR_ENABLED = ON
2023-05-09 07:45:46.271 : nlstddp_dump_ptable:  TRACE_LEVEL_CLIENT = USER
2023-05-09 07:45:46.271 : nlstddp_dump_ptable:  WALLET_LOCATION = (SOURCE = (METHOD = FILE) (METHOD_DATA = (DIRECTORY = /opt/shared/shsdb/operation/etc/oracle_wallet)))
2023-05-09 07:45:46.271 : nlstddp_dump_ptable:  ADR_BASE = /tmp/mylogs
2023-05-09 07:45:46.271 : nlstddp_dump_ptable:  SQLNET.WALLET_OVERRIDE = TRUE
2023-05-09 07:45:46.271 : nlstdtp_trace_pfile:--- PARAMETER SOURCE INFORMATION ENDS ---
2023-05-09 07:45:46.271 : nlstddl_do_alter_log:--- LOG CONFIGURATION INFORMATION FOLLOWS ---
2023-05-09 07:45:46.271 : nlstddl_do_alter_log:Log stream will be "/tmp/mylogs/oradiag_te504ho2/diag/clients/user_te504ho2/host_885234334_110/alert/log.xml"
2023-05-09 07:45:46.271 : nlstddl_do_alter_log:Log stream validation not requested
2023-05-09 07:45:46.272 : nlstddl_do_alter_log:--- LOG CONFIGURATION INFORMATION ENDS ---

2023-05-09 07:45:46.284 : nrigbni:Unable to get data from navigation file tnsnav.ora
2023-05-09 07:45:46.672 : nigconcbs:Can't Get Oracle SID Name
```
There are no further lines after that, so no connect descriptor is shown.

**Comment by anthony-tuininga** [comment]
@glM26, the level of trace provided is insufficient to figure out the issue. Can you change the line in `sqlnet.ora` to say `trace_level_client=support` and rerun? Thanks!

**Comment by anthony-tuininga** [comment]
I have pushed a patch that should correct this issue and a relevant test case. If you are able to [build from source](https://python-oracledb.readthedocs.io/en/latest/user_guide/installation.html#installing-from-source-code) you can verify that it corrects your issue as well.

Someone else noted that if you do this it will also work (without the patch):

```
conn = oracledb.connect(dsn="@tns_alias")
```

**Comment by glM26** [comment]
@anthony-tuininga thank you for help, I can confirm that both suggestions (building the newest version from source as well as adding @ in front of the dsn) solved my issue.
Do you know when a new version of oracledb containing the fix will be released?

**Comment by anthony-tuininga** [comment]
Thanks, @glM26. We will discuss internally but generally we like to make a release that contains bug fixes within a few weeks at most. I will post here when a release is made that contains this patch.

**Comment by anthony-tuininga** [comment]
This has been fixed in python-oracledb 1.3.2 which was just released.
